File: examples/go/main-urlencoded.go
package main

/**
 * iPaymu Callback Handler - Go (URL-encoded)
 *
 * Server berjalan di port 8080
 * Format: application/x-www-form-urlencoded
 *
 * Cara menjalankan:
 *   go run main-urlencoded.go
 */

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
)

func getEnv(key, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}
	return fallback
}

var secretKey = getEnv("SECRET_KEY", "123456")

func normalizeData(data map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})

	for key, val := range data {
		switch key {
		case "is_escrow":
			switch v := val.(type) {
			case string:
				if v == "1" || strings.ToLower(v) == "true" {
					result[key] = 1
				} else {
					result[key] = 0
				}
			case bool:
				if v {
					result[key] = 1
				} else {
					result[key] = 0
				}
			case float64, int, int64:
				result[key] = v
			default:
				result[key] = 0
			}
		case "trx_id", "status_code", "transaction_status_code", "paid_off", "sub_total", "total", "amount", "fee":
			switch v := val.(type) {
			case string:
				if i, err := strconv.Atoi(v); err == nil {
					result[key] = i
				} else {
					result[key] = val
				}
			case float64:
				result[key] = int(v)
			default:
				result[key] = val
			}
		case "additional_info":
			switch v := val.(type) {
			case string:
				if v == "[]" || v == "" {
					result[key] = []interface{}{}
				} else {
					var arr []interface{}
					if err := json.Unmarshal([]byte(v), &arr); err == nil {
						result[key] = arr
					} else {
						result[key] = []interface{}{}
					}
				}
			case []interface{}:
				result[key] = v
			default:
				result[key] = []interface{}{}
			}
		default:
			result[key] = fmt.Sprintf("%v", val)
		}
	}

	if _, ok := result["additional_info"]; !ok {
		result["additional_info"] = []interface{}{}
	}

	return result
}

func phpKsort(data map[string]interface{}) map[string]interface{} {
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	sorted := make(map[string]interface{})
	for _, k := range keys {
		sorted[k] = data[k]
	}

	return sorted
}

func generateSignature(data map[string]interface{}) string {
	normalizedData := normalizeData(data)
	if _, ok := normalizedData["signature"]; ok {
		delete(normalizedData, "signature")
	}

	sortedData := phpKsort(normalizedData)
	jsonBytes, _ := json.Marshal(sortedData)
	jsonBody := string(jsonBytes)
	jsonBody = strings.ReplaceAll(jsonBody, "/", "\\/")

	h := hmac.New(sha256.New, []byte(secretKey))
	h.Write([]byte(jsonBody))

	return hex.EncodeToString(h.Sum(nil))
}

func callbackHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, `{"status":"error","message":"Method Not Allowed"}`, http.StatusMethodNotAllowed)
		return
	}

	if err := r.ParseForm(); err != nil {
		http.Error(w, `{"status":"error","message":"Invalid form data"}`, http.StatusBadRequest)
		return
	}

	receivedSignature := r.Header.Get("X-Signature")
	if receivedSignature == "" {
		receivedSignature = r.FormValue("signature")
	}

	if receivedSignature == "" {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{
			"status":  "error",
			"message": "Missing X-Signature header",
		})
		return
	}

	data := make(map[string]interface{})
	for key, values := range r.Form {
		if len(values) > 0 {
			data[key] = values[0]
		}
	}

	calculatedSignature := generateSignature(data)

	w.Header().Set("Content-Type", "application/json")
	if hmac.Equal([]byte(calculatedSignature), []byte(receivedSignature)) {
		fmt.Println("✅ Signature valid!")
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]string{
			"status":  "OK",
			"message": "Callback processed successfully",
		})
	} else {
		fmt.Printf("❌ Invalid signature! Expected: %s, Got: %s\n", receivedSignature, calculatedSignature)
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]interface{}{
			"status":  "error",
			"message": "Invalid Signature",
			"debug": map[string]string{
				"received":   receivedSignature,
				"calculated": calculatedSignature,
			},
		})
	}
}

func main() {
	port := getEnv("PORT", "8080")

	http.HandleFunc("/callback", callbackHandler)

	fmt.Println("═══════════════════════════════════════════════════")
	fmt.Println("  iPaymu Callback Handler - URL-encoded (Go)")
	fmt.Println("═══════════════════════════════════════════════════")
	fmt.Printf("  Server running on http://localhost:%s\n", port)
	fmt.Printf("  Endpoint: POST http://localhost:%s/callback\n", port)
	fmt.Println("═══════════════════════════════════════════════════")

	if err := http.ListenAndServe(":"+port, nil); err != nil {
		fmt.Printf("Server error: %v\n", err)
	}
}

File: examples/rust/src/main-urlencoded.rs
use actix_web::{web, App, HttpResponse, HttpServer, post, HttpRequest};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use serde_json::{Value, Map};
use std::collections::BTreeMap;

// Tipe untuk HMAC-SHA256
type HmacSha256 = Hmac<Sha256>;

const SECRET_KEY: &str = "123456"; // Nomor VA akun iPaymu Anda (lihat di Dashboard > Integration > API Key)

// Normalisasi data dari form parameter
// URL Encoded Forms akan mengirimkan semuanya sebagai string.
fn normalize_data(data: &Map<String, Value>) -> Map<String, Value> {
    let mut result = Map::new();
    
    for (key, val) in data.iter() {
        let normalized_val = match key.as_str() {
            "is_escrow" => {
                match val {
                    Value::String(s) => {
                        let is_true = s == "1" || s == "true";
                        Value::Number((if is_true { 1 } else { 0 }).into())
                    },
                    Value::Bool(b) => Value::Number((if *b { 1 } else { 0 }).into()),
                    Value::Number(n) => Value::Number(n.clone()),
                    _ => Value::Number(0.into()),
                }
            }
            "trx_id" | "status_code" | "transaction_status_code" | "paid_off" | "sub_total" | "total" | "amount" | "fee" => {
                match val {
                    Value::String(s) => {
                        s.parse::<i64>().map(|n| Value::Number(n.into()))
                            .unwrap_or_else(|_| val.clone())
                    }
                    Value::Number(n) => Value::Number(n.clone()),
                    _ => val.clone(),
                }
            }
            "additional_info" => {
                match val {
                    Value::String(s) if s == "[]" => Value::Array(vec![]),
                    Value::Array(a) => Value::Array(a.clone()),
                    _ => Value::Array(vec![]),
                }
            }
            _ => Value::String(val.as_str().unwrap_or("").to_string()),
        };
        
        result.insert(key.clone(), normalized_val);
    }
    
    // Pastikan additional_info ada
    if !result.contains_key("additional_info") {
        result.insert("additional_info".to_string(), Value::Array(vec![]));
    }
    
    result
}

// Urutkan key sesuai PHP ksort
fn php_ksort(data: &Map<String, Value>) -> BTreeMap<String, Value> {
    let mut sorted = BTreeMap::new();
    for (k, v) in data.iter() {
        // Abaikan signature dalam proses pembuatan signature
        if k != "signature" {
            sorted.insert(k.clone(), v.clone());
        }
    }
    sorted
}

fn generate_signature(data: &Map<String, Value>) -> String {
    // Sort key
    let sorted = php_ksort(data);
    
    // Convert to JSON
    let json_body = serde_json::to_string(&sorted).unwrap();
    
    // Escape slashes seperti PHP
    let json_body = json_body.replace("/", "\\/");
    println!("DEBUG JSON BODY: {}", json_body);
    
    // Generate HMAC-SHA256
    let mut mac = HmacSha256::new_from_slice(SECRET_KEY.as_bytes())
        .expect("HMAC can take key of any size");
    mac.update(json_body.as_bytes());
    
    hex::encode(mac.finalize().into_bytes())
}

#[post("/callback")]
async fn callback_handler(req: HttpRequest, body: web::Form<std::collections::HashMap<String, String>>) -> HttpResponse {
    // Ambil signature dari header
    let received_sig = req.headers()
        .get("X-Signature")
        .and_then(|h| h.to_str().ok())
        .unwrap_or("");
    
    // Convert HashMap<String, String> ke Map<String, Value>
    let mut data = Map::new();
    for (k, v) in body.into_inner() {
        data.insert(k, Value::String(v));
    }
    
    println!("DEBUG RAW FORM MAP: {:?}", data);
    
    // Lakukan normalisasi (string menjadi format yg benar)
    let normalized = normalize_data(&data);
    println!("DEBUG NORMALIZED MAP: {:?}", normalized);

    // Generate signature untuk divalidasi
    let calculated_sig = generate_signature(&normalized);
    println!("Received Signature: {}", received_sig);
    println!("Calculated Signature: {}", calculated_sig);
    
    // Validasi
    if calculated_sig == received_sig {
        println!("✅ Signature valid");
        // Lakukan proses bisnis terkait status transaksi
        HttpResponse::Ok().json(serde_json::json!({
            "status": "OK"
        }))
    } else {
        println!("❌ Signature tidak valid");
        HttpResponse::BadRequest().json(serde_json::json!({
            "status": "error",
            "message": "Invalid Signature"
        }))
    }
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    println!("Server running on :8082");
    // Port 8082 untuk x-www-form-urlencoded
    HttpServer::new(|| {
        App::new().service(callback_handler)
    })
    .bind("127.0.0.1:8082")?
    .run()
    .await
}
